// @ngx-cache/core/esm2015/cache.module.js
import { InjectionToken, Injector, NgModule, Optional, PLATFORM_ID, SkipSelf } from '@angular/core';
import { CacheLoader, CacheStaticLoader } from './cache.loader';
import { CacheService } from './cache.service';
import * as i0 from "@angular/core";
export const STORAGE = new InjectionToken('STORAGE');
export const cacheFactory = () => new CacheStaticLoader();
export const cacheServiceFactory = (loader, platformId, injector) => new CacheService(loader, platformId, injector);
export class CacheModule {
    constructor(parentModule) {
        if (parentModule) {
            throw new Error('CacheModule already loaded; import in root module only.');
        }
    }
    static forRoot(configuredProvider = {
        provide: CacheLoader,
        useFactory: cacheFactory
    }) {
        return {
            ngModule: CacheModule,
            providers: [
                configuredProvider,
                {
                    provide: CacheService,
                    useFactory: cacheServiceFactory,
                    deps: [CacheLoader, PLATFORM_ID, Injector]
                }
            ]
        };
    }
}
CacheModule.ɵmod = i0.ɵɵdefineNgModule({ type: CacheModule });
CacheModule.ɵinj = i0.ɵɵdefineInjector({ factory: function CacheModule_Factory(t) { return new (t || CacheModule)(i0.ɵɵinject(CacheModule, 12)); } });
(function () { i0.ɵsetClassMetadata(CacheModule, [{
        type: NgModule
    }], function () { return [{ type: CacheModule, decorators: [{
                type: Optional
            }, {
                type: SkipSelf
            }] }]; }, null); })();
