// @ngx-cache/core/esm5/cache.module.js
import { InjectionToken, Injector, NgModule, Optional, PLATFORM_ID, SkipSelf } from '@angular/core';
import { CacheLoader, CacheStaticLoader } from './cache.loader';
import { CacheService } from './cache.service';
import * as i0 from "@angular/core";
export var STORAGE = new InjectionToken('STORAGE');
export var cacheFactory = function () { return new CacheStaticLoader(); };
export var cacheServiceFactory = function (loader, platformId, injector) {
    return new CacheService(loader, platformId, injector);
};
var CacheModule = (function () {
    function CacheModule(parentModule) {
        if (parentModule) {
            throw new Error('CacheModule already loaded; import in root module only.');
        }
    }
    CacheModule.forRoot = function (configuredProvider) {
        if (configuredProvider === void 0) { configuredProvider = {
            provide: CacheLoader,
            useFactory: cacheFactory
        }; }
        return {
            ngModule: CacheModule,
            providers: [
                configuredProvider,
                {
                    provide: CacheService,
                    useFactory: cacheServiceFactory,
                    deps: [CacheLoader, PLATFORM_ID, Injector]
                }
            ]
        };
    };
    CacheModule.ɵmod = i0.ɵɵdefineNgModule({ type: CacheModule });
    CacheModule.ɵinj = i0.ɵɵdefineInjector({ factory: function CacheModule_Factory(t) { return new (t || CacheModule)(i0.ɵɵinject(CacheModule, 12)); } });
    return CacheModule;
}());
export { CacheModule };
(function () { i0.ɵsetClassMetadata(CacheModule, [{
        type: NgModule
    }], function () { return [{ type: CacheModule, decorators: [{
                type: Optional
            }, {
                type: SkipSelf
            }] }]; }, null); })();
